src/Modules/Trinity.DynamicCluster/Trinity.DynamicCluster/DynamicMemoryCloud/DynamicMemoryCloud.cs

```diff
@@ -0,0 +1,233 @@
+                         .First();
+            action(temp_id);
+            temporaryRemoteStorageRepo.TryRemove(temp_id, out var _);
+        }
+
+        private void CheckServerProtocolSignatures(DynamicRemoteStorage rs)
+        {
+            Log.WriteLine(LogLevel.Debug, $"Checking protocol signatures with '{rs.NickName}' ({rs.Id})...");
+
+            CheckProtocolSignatures_impl(rs, cluster_config.RunningMode, RunningMode.Server);
+        }
+
+        internal TrinityErrorCode OnStorageJoin(DynamicRemoteStorage remoteStorage)
+        {
+            _DoWithTempStorage(remoteStorage, id =>
+            {
+                var module = GetCommunicationModule<DynamicClusterCommModule>();
+                using (var rsp = module.QueryChunkedRemoteStorageInformation(id))
+                {
+                    remoteStorage.SetId(rsp.info.id, rsp.info.partition);
+                    ChunkedStorageTable(rsp.info.partition).Mount(remoteStorage, rsp);
+                }
+                CheckServerProtocolSignatures(remoteStorage);
+                Log.WriteLine($"DynamicCluster: Connected to '{remoteStorage.NickName}' ({remoteStorage.Id})");
+            });
+            return TrinityErrorCode.E_SUCCESS;
+        }
+
+        /// <summary>
+        /// It is guaranteed that OnStorageLeave will only be called
+        /// on those storages that has been previously sent to
+        /// <see cref="OnStorageJoin"/>. However, OnStorageLeave may
+        /// be called multiple times (disconnected, then reconnected,
+        /// or remote <see cref="Close"/> is called).
+        /// </summary>
+        /// <param name="remoteStorage"></param>
+        internal TrinityErrorCode OnStorageLeave(int partition, Guid Id)
+        {
+            var module = GetCommunicationModule<DynamicClusterCommModule>();
+            var remote_storage = ChunkedStorageTable(partition)
+                                .OfType<DynamicRemoteStorage>()
+                                .FirstOrDefault(_ => _.Id == Id);
+            return ChunkedStorageTable(partition).Unmount(remote_storage);
+        }
+
+        public string NickName { get; private set; }
+
+        public override IEnumerable<Chunk> MyChunks => m_partitioner.MyChunkList;
+        public override int MyPartitionId => m_partitioner.PartitionId;
+        public override int PartitionCount => m_partitioner.PartitionCount;
+        public int ChunkCount => m_partitioner.ChunkCount;
+        public Guid InstanceId => m_nameservice.InstanceId;
+        public override bool IsLocalCell(long cellId)
+        {
+            return (GetStorageByCellId(cellId) as Partition).IsLocal(cellId);
+        }
+        public override int MyProxyId => -1;
+        public override int ProxyCount => 0;
+        public override bool Open(ClusterConfig config, bool nonblocking)
+        {
+            InitializeComponents();
+
+            m_partitioner.Start(m_cancelSrc.Token);
```

Comment:
note, implementations should make sure that components do not have dependencies on each other.. or, do we need to implement dependency check (like AutoFac)?